Redesign translator post processor and adding custom dictionary feature to translator module

amr-elsehemy opened the pull request:
- Redesigned translator post processor to be more extensible and decoupled from other components logic 

- Added custom dictionary feature to the translator module, so that users can pass to the translator middle ware a custom vocab dictionary to forcibly override the translation output for some words

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/botbuilder-dotnet?pullRequest=600) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/botbuilder-dotnet?pullRequest=600) before we can accept your contribution.<br/><br/>:x: amr-elsehemy [sign now](https://cla.opensource.microsoft.com/Microsoft/botbuilder-dotnet?pullRequest=600)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/botbuilder-dotnet?pullRequest=600) it.</sub>

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PatternsPostProcessor.cs:
```diff
@@ -0,0 +1,127 @@
+        /// <param name="translatedDocument">Translated document</param>
+        /// <param name="currentLanguage">Current source language</param>
+        /// <returns>A Task represents the asynchronus operation</returns>
+        public PostProcessedDocument Process(TranslatedDocument translatedDocument, string currentLanguage)
```
Null check the passed in arguments, and throw if they're null.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PatternsPostProcessor.cs:
```diff
@@ -0,0 +1,127 @@
+        /// <returns>A Task represents the asynchronus operation</returns>
+        public PostProcessedDocument Process(TranslatedDocument translatedDocument, string currentLanguage)
+        {
+            bool containsNum = Regex.IsMatch(translatedDocument.SourceMessage, @"\d");
```
This will throw an NRE if a null is passed in.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PatternsPostProcessor.cs:
```diff
@@ -0,0 +1,127 @@
+            }
+            if (temporaryPatterns.Count == 0 && !containsNum)
+                processedResult = translatedDocument.TargetMessage;
+            if (string.IsNullOrWhiteSpace(translatedDocument.RawAlignment))
```
whitespace.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PatternsPostProcessor.cs:
```diff
@@ -0,0 +1,127 @@
+            if (string.IsNullOrWhiteSpace(translatedDocument.RawAlignment))
+                processedResult = translatedDocument.TargetMessage;
+
+            var toBeReplaced = from result in temporaryPatterns
```
I've tried really hard to avoid using LINQ statements like this. Is there a more clear way you could write this w/o LINQ? 

My experience (and the teams) is that these statements can be very hard to debug over time.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PatternsPostProcessor.cs:
```diff
@@ -0,0 +1,127 @@
+                foreach (string pattern in toBeReplaced)
+                {
+                    Match matchNoTranslate = Regex.Match(translatedDocument.SourceMessage, pattern, RegexOptions.Singleline | RegexOptions.IgnoreCase);
+                    int noTranslateStartChrIndex = matchNoTranslate.Groups[1].Index;
```
Comments and examples. I have no idea what this code is doing, and I don't have the time to step through it in a debugger. The more clear you can make it, with comments + examples, the better.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessedDocument.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
//Copyright

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+        /// <param name="words">String Array of words to be joined.</param> 
+        /// <returns>string joined sentence</returns>
+        public static string Join(string delimiter, string[] words)
+        {
```
null checks

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+
+namespace Microsoft.Bot.Builder.Ai.Translation.PostProcessor
+{
+    internal class PostProcessingUtilities
```
Make this public. Then write unit tests against it. :) 

Internal classes are hard to tests, and I do notice that there are no tests at all against this code.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+        /// </summary>
+        /// <param name="sentence">String containing sentence to be splitted.</param> 
+        /// <returns>string array of words.</returns>
+        public static string[] SplitSentence(string sentence, string[] alignments = null, bool isSrcSentence = true)
```
This method needs a fully pathalogical set of tests.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+                }
+                string withoutSpaceSentence = sentence.Replace(" ", "");
+
+                foreach (string alignData in alignments)
```
This is scary code, as I have no idea what it's doing and no means to judge if it's correct. 

Comments + Sampels, and lots of tests!!

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+        }
+
+        /// <summary>
+        ///parsing alignment information onto a dictionary
```
whitespace

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+        /// </summary>
+        /// <param name="alignment">String containing phrase alignments</param>
+        /// <param name="sourceMessage">String containing source message</param>
+        /// /<param name="trgMessage">String containing translated message</param>
```
exatra "/"

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+        /// <returns></returns>
+        public static Dictionary<int, int> WordAlignmentParse(string[] alignments, string[] srcWords, string[] trgWords)
+        {
+            Dictionary<int, int> alignMap = new Dictionary<int, int>();
```
argument checking.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+            Dictionary<int, int> alignMap = new Dictionary<int, int>();
+            string sourceMessage = Join(" ", srcWords);
+            string trgMessage = Join(" ", trgWords);
+            foreach (string alignData in alignments)
```
More scary code. Please simplify, comment, examples, and tests.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PostProcessingUtilities.cs:
```diff
@@ -0,0 +1,126 @@
+        /// <param name="target">Target Language</param>
+        /// <param name="srcWrd">Source Word</param>
+        /// <returns></returns>
+        public static string[] KeepSrcWrdInTranslation(Dictionary<int, int> alignment, string[] sourceWords, string[] targetWords, int srcWrdIndx)
```
KeepSourceWordInTranslation

cleemullins replied:
Argument checking.

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/TranslatedDocument.cs:
```diff
@@ -0,0 +1,48 @@
+﻿using System;
```
copyright

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/TranslatedDocument.cs:
```diff
@@ -0,0 +1,48 @@
+    /// </summary>
+    public class TranslatedDocument
+    {
+        private string sourceMessage;
```
member variables should have a "_" on them.

cleemullins replied:
Most of these variables are not needed. They can be "implied" form the properties. 

public string SourceMessage { get ;set; }

cleemullins commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/TranslatedDocument.cs:
```diff
@@ -0,0 +1,48 @@
+        /// <param name="targetMessage">Target/translated message</param>
+        public TranslatedDocument(string sourceMessage, string targetMessage)
+        {
+            this.sourceMessage = sourceMessage;
```
if (string.IsNullOrWhitespace(sourceMessage)) throw new ARgumentNullException(nameof(...))

cleemullins commented on tests/Microsoft.Bot.Builder.Ai.Translation.Tests/TranslatorTests.cs:
```diff
@@ -69,18 +74,111 @@ public class TranslatorTests
 
             Translator translator = new Translator(translatorKey);
-            translator.SetPostProcessorTemplate(new List<string> { "perr[oa]" });
+            Dictionary<string, List<string>> patterns = new Dictionary<string, List<string>>();
```
This entire test needs to be broken down into a larger number of simpler tests. Those tests need to focus on boundry condidtions, passing in null variables, passing in empty collections, and the like. Once *those* tests pass, the actual functionality needs to be tested.

cleemullins commented on tests/Microsoft.Bot.Builder.Ai.Translation.Tests/TranslatorTests.cs:
```diff
@@ -69,18 +74,111 @@ public class TranslatorTests
+        [TestCategory("Translator")]
+        public async Task Translator_DictionaryTest()
+        {
+            if (!EnvironmentVariablesDefined())
```
This test is too complex. Please decompose into many smaller, simpler, tests.

cleemullins commented on tests/Microsoft.Bot.Builder.Ai.Translation.Tests/TranslatorTests.cs:
```diff
@@ -69,18 +74,111 @@ public class TranslatorTests
+        [TestCategory("Translator")]
+        public async Task Translator_PatternsAndDictionaryTest()
+        {
+            if (!EnvironmentVariablesDefined())
```
Same here. Tooo complex. Please decompose into actual unit tests.

amr-elsehemy commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/IPostProcessor.cs:
```diff
@@ -0,0 +1,20 @@
+        /// <param name="translatedDocument">Translated document</param>
+        /// <param name="currentLanguage">Current source language</param>
+        /// <returns></returns>
+        PostProcessedDocument Process(TranslatedDocument translatedDocument, string currentLanguage);
```
Actually we have both language and locale in our definitions for ex : 
Language = "en" 
Locale = "en-us"
so I guess language would be more appropriate to use than locale

amr-elsehemy commented on libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/CustomDictionaryPostProcessor.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            if (_userCustomDictionaries[currentLanguage].Count > 0)
+            {
+                for (int i = 0; i < translatedDocument.SourceTokens.Length; i++)
```
I preferred this for loop format because i need to use the index in the loop body, so instead of using a separate counter with a  foreach I used this format .

cleemullins commented:
@amr-elsehemy - Any updates here? I would love to see this cleaned up so we can take it.

amr-elsehemy commented:
@cleemullins, sure Chris, I'm almost finished, I did the following 

- introduced a better Custom dictionary data structure
- added parameter validations everywhere that was missing
- written unit tests for the PostProcessorUtilities
-simplified patterns post processor and got rid of the linq

I'm just working on enriching the comments and examples and will finish soon.

amr-elsehemy commented:
@cleemullins - I've added my new updates I tried to consider all your comments and suggestions as much as possible, please review the current state and give me your feedback :))

amr-elsehemy commented:
@cleemullins , 
kind reminder to review my updates :)

cleemullins commented:
@amr-elsehemy I'll got through this today.
